modules/140-user-authz/hooks/handle_manage_bindings.go

```diff
@@ -114,135 +105,116 @@ func filterManageBinding(obj *unstructured.Unstructured) (go_hook.FilterResult,
 }
 
 type filteredManageRole struct {
-	Name  string `json:"name"`
-	Level string `json:"level"`
-	// manage fields
-	Scope     string `json:"scope"`
-	Namespace string `json:"namespace"`
-	// module cap fields
-	Namespaces []string `json:"namespaces"`
-	Scopes     []string `json:"scopes"`
+	Name   string                  `json:"name"`
+	Labels map[string]string       `json:"aggregationLabels"`
+	Rule   *rbacv1.AggregationRule `json:"selectors"`
 }
 
 func filterManageRole(obj *unstructured.Unstructured) (go_hook.FilterResult, error) {
 	var role rbacv1.ClusterRole
 	if err := sdk.FromUnstructured(obj, &role); err != nil {
 		return nil, err
 	}
-	filtered := &filteredManageRole{
-		Name: role.Name,
-	}
-	for key, val := range role.Labels {
-		switch key {
-		case "rbac.deckhouse.io/namespace":
-			filtered.Namespace = val
-		case "rbac.deckhouse.io/scope":
-			filtered.Scope = val
-		case "rbac.deckhouse.io/level":
-			filtered.Level = val
-		}
-		if involved, ok := strings.CutPrefix(key, "rbac.deckhouse.io/aggregate-to-"); ok {
-			involved, _ = strings.CutSuffix(involved, "-as")
-			filtered.Scopes = append(filtered.Scopes, involved)
-		}
-	}
-	if filtered.Level == "all" {
-		filtered.Scope = "all"
-	}
-	return filtered, nil
+	return &filteredManageRole{
+		Name:   role.Name,
+		Labels: role.Labels,
+		Rule:   role.AggregationRule,
+	}, nil
 }
 
-func syncRoles(input *go_hook.HookInput) error {
-	roles := parseRoles(input.Snapshots["manageRoles"])
-	expected := make(map[string]*filteredUseBinding)
-	for _, snapBinding := range input.Snapshots["manageBindings"] {
-		binding := snapBinding.(*filteredManageBinding)
-		namespaces, ok := roles[binding.RoleName]
-		if !ok {
-			continue
-		}
-		splits := strings.Split(binding.RoleName, ":")
-		for _, namespace := range namespaces {
-			expectedBinding := &filteredUseBinding{
-				Name:        fmt.Sprintf("d8:use:binding:%s", binding.Name),
-				Namespace:   namespace,
-				RelatedWith: binding.Name,
-				RoleName:    fmt.Sprintf("d8:use:role:%s", splits[len(splits)-1]),
-				Subjects:    binding.Subjects,
-			}
-			input.PatchCollector.Create(createBinding(expectedBinding), object_patch.UpdateIfExists())
-			expected[expectedBinding.Name] = expectedBinding
+func syncBindings(input *go_hook.HookInput) error {
+	expected := make(map[string]bool)
+	for _, snap := range input.Snapshots["manageBindings"] {
+		role, namespaces := roleAndNamespacesByBinding(input.Snapshots["manageRoles"], snap.(*filteredManageBinding).RoleName)
```

Comment:
you're cast 3 times here. cast once and use result
```suggestion
fmb := snap.(*filteredManageBinding)
		role, namespaces := roleAndNamespacesByBinding(input.Snapshots["manageRoles"], fmb.RoleName)
```

Reply:
done